gstd/gstd_http.c

```diff
@@ -0,0 +1,476 @@
+{
+  gchar *response = NULL;
+  gchar *message = NULL;
+  SoupURI *address = NULL;
+  GstdReturnCode ret = GSTD_EOK;
+  gchar *output = NULL;
+  const gchar *description = NULL;
+  SoupStatus status = SOUP_STATUS_OK;
+
+  g_return_if_fail (server);
+  g_return_if_fail (msg);
+  g_return_if_fail (session);
+
+  address = soup_message_get_uri (msg);
+
+  message = g_strdup_printf ("read %s", soup_uri_get_path (address));
+  ret = gstd_parser_parse_cmd (session, message, &output);      // in the parser
+  // Prepend the code to the output
+  description = gstd_return_code_to_string (ret);
+  response =
+      g_strdup_printf
+      ("{\n  \"code\" : %d,\n  \"description\" : \"%s\",\n  \"response\" : %s\n}",
+      ret, description, output ? output : "null");
+
+  soup_message_set_response (msg, "application/json", SOUP_MEMORY_COPY,
+      response, strlen (response));
+
+  status = get_status_code (ret);
+  soup_message_set_status (msg, status);
+
+  g_free (output);
+  g_free (response);
+  g_free (message);
+
+  return;
+}
+
+static void
+do_post (SoupServer * server, SoupMessage * msg, GHashTable * query,
+    GstdSession * session)
+{
+  gchar *response = NULL;
+  gchar *message = NULL;
+  gchar *name = NULL;
+  gchar *description_pipe = NULL;
+  SoupURI *address = NULL;
+  GstdReturnCode ret = GSTD_EOK;
+  gchar *output = NULL;
+  const gchar *description = NULL;
+  SoupStatus status = SOUP_STATUS_OK;
+  const gchar *query_text = NULL;
+
+  g_return_if_fail (server);
+  g_return_if_fail (msg);
+  g_return_if_fail (query);
+  g_return_if_fail (session);
+
+  address = soup_message_get_uri (msg);
+
+  query_text = soup_uri_get_query (address);
+  if (query_text) {
+    query = soup_form_decode (query_text);
+  } else {
+    ret = GSTD_BAD_VALUE;
+    GST_INFO_OBJECT (session, "No query params provided");
+    goto out;
+  }
+
+  name = g_hash_table_lookup (query, "name");
+  if (!name) {
+    ret = GSTD_BAD_VALUE;
+    GST_INFO_OBJECT (session,
+        "Wrong query param provided, \"name\" doesn't exist");
+    goto out;
+  }
+  description_pipe = g_hash_table_lookup (query, "description");
+  if (!description_pipe) {
+    ret = GSTD_BAD_VALUE;
+    GST_INFO_OBJECT (session,
+        "Wrong query param provided, \"description\" doesn't exist");
+    goto out;
+  }
+  message = g_strdup_printf
+      ("create %s %s %s", soup_uri_get_path (address), name, description_pipe);
+  ret = gstd_parser_parse_cmd (session, message, &output);      // in the parser
+  // Prepend the code to the output
+  description = gstd_return_code_to_string (ret);
+  response =
+      g_strdup_printf
+      ("{\n  \"code\" : %d,\n  \"description\" : \"%s\",\n  \"response\" : %s\n}",
+      ret, description, output ? output : "null");
+
+  soup_message_set_response (msg, "application/json", SOUP_MEMORY_COPY,
+      response, strlen (response));
+out:
+  status = get_status_code (ret);
+  soup_message_set_status (msg, status);
```

Comment:
Set the status before setting the response.

> To set the response status, call soup_message_set_status or soup_message_set_status_full
From: [libsoup-server-howto](https://developer.gnome.org/libsoup/stable/libsoup-server-howto.html)

Reply:
Why this order?